node/node.go

```diff
@@ -282,50 +279,48 @@ func (node *Node) tryBroadcastStaking(stakingTx *staking.StakingTransaction) {
 
 // Add new transactions to the pending transaction list.
 func (node *Node) addPendingTransactions(newTxs types.Transactions) {
-	node.TxPool.AddRemotes(newTxs)
+	poolTxs := types.PoolTransactions{}
+	for _, tx := range newTxs {
+		poolTxs = append(poolTxs, tx)
+	}
+	node.TxPool.AddRemotes(poolTxs)
 
 	pendingCount, queueCount := node.TxPool.Stats()
-	utils.Logger().Info().Int("length of newTxs", len(newTxs)).Int("totalPending", pendingCount).Int("totalQueued", queueCount).Msg("Got more transactions")
+	utils.Logger().Info().
+		Int("length of newTxs", len(newTxs)).
+		Int("totalPending", pendingCount).
+		Int("totalQueued", queueCount).
+		Msg("Got more transactions")
 }
 
 // Add new staking transactions to the pending staking transaction list.
 func (node *Node) addPendingStakingTransactions(newStakingTxs staking.StakingTransactions) {
-	// TODO: incorporate staking txn into TxPool
 	if node.NodeConfig.ShardID == shard.BeaconChainShardID &&
 		node.Blockchain().Config().IsPreStaking(node.Blockchain().CurrentHeader().Epoch()) {
-		node.pendingStakingTxMutex.Lock()
+		poolTxs := types.PoolTransactions{}
 		for _, tx := range newStakingTxs {
-			const txPoolLimit = 1000
-			if s := len(node.pendingStakingTransactions); s >= txPoolLimit {
-				utils.Logger().Info().
-					Int("tx-pool-size", s).
-					Int("tx-pool-limit", txPoolLimit).
-					Msg("Current staking txn pool reached limit")
-				break
-			}
-			if _, ok := node.pendingStakingTransactions[tx.Hash()]; !ok {
-				node.pendingStakingTransactions[tx.Hash()] = tx
-			}
+			poolTxs = append(poolTxs, tx)
 		}
+		node.TxPool.AddRemotes(poolTxs)
+		pendingCount, queueCount := node.TxPool.Stats()
 		utils.Logger().Info().
-			Int("length of newStakingTxs", len(newStakingTxs)).
-			Int("totalPending", len(node.pendingStakingTransactions)).
+			Int("length of newStakingTxs", len(poolTxs)).
+			Int("totalPending", pendingCount).
+			Int("totalQueued", queueCount).
 			Msg("Got more staking transactions")
-		node.pendingStakingTxMutex.Unlock()
 	}
 }
 
 // AddPendingStakingTransaction staking transactions
-func (node *Node) AddPendingStakingTransaction(
-	newStakingTx *staking.StakingTransaction) {
-	// TODO: everyone should record staking txns, not just leader
-	if node.Consensus.IsLeader() &&
-		node.NodeConfig.ShardID == shard.BeaconChainShardID {
+func (node *Node) AddPendingStakingTransaction(newStakingTx *staking.StakingTransaction) {
+	if node.NodeConfig.ShardID == shard.BeaconChainShardID {
 		node.addPendingStakingTransactions(staking.StakingTransactions{newStakingTx})
-	} else {
-		utils.Logger().Info().Str("Hash", newStakingTx.Hash().Hex()).Msg("Broadcasting Staking Tx")
-		node.tryBroadcastStaking(newStakingTx)
+		if node.NodeConfig.Role() != nodeconfig.ExplorerNode {
```

Comment:
The original logic is still correct, we want any node that's not the leader to broadcast the staking transaction (not just the explorer node, it's a single point of failure)

Reply:
discussed in person, will also change regular tx logic so that everyone broadcasts in this PR